golang/src/stream_reader.rs

```diff
@@ -0,0 +1,167 @@
+use pravega_client::event::reader::{Event, EventReader, EventReaderError, SegmentSlice};
+use pravega_client_shared::ScopedStream;
+use std::panic::{catch_unwind, AssertUnwindSafe};
+use std::ptr;
+use tokio::runtime::Handle;
+use crate::error::{clear_error, set_error};
+use crate::memory::{Buffer, set_buffer};
+use tokio::sync::mpsc::UnboundedSender;
+use crate::reactor::*;
+pub struct StreamReader {
+    reader: EventReader,
+    runtime_handle: Handle,
+    streams: Vec<ScopedStream>,
+    sender: UnboundedSender<Incoming>
+}
+
+impl StreamReader {
+    pub fn new(reader: EventReader, runtime_handle: Handle, streams: Vec<ScopedStream>, sender: UnboundedSender<Incoming>) -> Self {
+        StreamReader {
+            reader,
+            runtime_handle,
+            streams,
+            sender
+        }
+    }
+
+    pub async fn get_segment_slice(&mut self) -> Result<Option<SegmentSlice>, EventReaderError> {
+        self.reader.acquire_segment().await
+    }
+
+    pub fn release_segment(&mut self, slice: Option<SegmentSlice>) -> Result<(), EventReaderError> {
+        if let Some(s) = slice {
+            self.runtime_handle.block_on(self.reader.release_segment(s))?;
+        }
+        Ok(())
+    }
+}
+
+pub struct SafeStreamReader {
+    v: *mut StreamReader
+}
+
+impl SafeStreamReader {
+    pub fn new(v: *mut StreamReader) -> Self {
+        Self { v }
+    }
+    pub fn get(&self) -> &mut StreamReader {
+        unsafe { &mut *(self.v) }
+    }
+}
+
+unsafe impl Send for SafeStreamReader {}
```

Comment:
As discussed, this is not safe. The compiler was correct to flag the error. We may need to either use an ID and a map, or find a way to pass ownership of a boxed reader. (Whichever ends up being easier)

Reply:
If my latest implement is accepted, this part won't be needed.